src/Storages/Statistics/ConditionSelectivityEstimator.cpp

```diff
@@ -124,7 +127,37 @@ bool ConditionSelectivityEstimator::extractAtomFromTree(const RPNBuilderTreeNode
 
         if (num_args == 2)
         {
-            if (func.getArgumentAt(1).tryGetConstant(const_value, const_type))
+            const bool is_in_operator = functionIsInOperator(func_name);
+
+            /// If the second argument is built from `ASTNode`, it should fall into next branch, which directly
+            /// extracts constant value from `ASTLiteral`. Otherwise we try to build `Set` from `ActionsDAG::Node`,
+            /// and extract constant value from it.
+            if (is_in_operator && !func.getArgumentAt(1).getASTNode())
```

Comment:
which cases are solved by this branch? Could you please write a unit test in `src/Storages/Statistics/tests/gtest_stats.cpp`?

Reply:
This branch was covered by `tests/queries/0_stateless/03580_improve_prewhere.sql`,  with sql: 
```
SELECT * FROM test_improve_prewhere
WHERE date = '2025-08-05' and lower(primary_key) in '00' and normal_column != 'hello' and value < 100
```

Reply:
If clickhouse are executing a real query with prewhere condition like `a in (xx)`, it would run into this branch because the second argument is `ActionsDAG::Node` instead of `IAST`. Without this branch, it would fall into next branch, which extract null value as `const_value` from `ColumnSet`, and it is not right.